websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java

```diff
@@ -44,21 +45,41 @@
 @MetaInfServices(Provider.class)
 public class Jetty10Provider implements Provider {
 
+    /**
+     * Number of seconds a WebsocketConnection may stay idle until it expires.
+     * Zero to disable.
+     * This value must be higher than the <code>jenkins.websocket.pingInterval</code>.
+     * Per <a href=https://www.eclipse.org/jetty/documentation/jetty-10/programming-guide/index.html#pg-websocket-session-ping>Jetty 10 documentation</a>
+     * a ping mechanism should keep the websocket active. Therefore, the idle timeout must be higher than the ping
+     * interval to avoid timeout issues.
+     */
+    private static long IDLE_TIMEOUT_SECONDS = Long.getLong("jenkins.websocket.idleTimeout", 60L);
+
     private static final String ATTR_LISTENER = Jetty10Provider.class.getName() + ".listener";
 
+    private JettyWebSocketServerContainer wsContainer;
```

Comment:
Caching this object is a change of behavior. Can I assume you tested repeated agent connections within a single controller session?

Reply:
usually this sort of configuration is done via `ServletContainerInitializer` mechanism.

Reply:
just an idea about those fields (same for `WebSocketServletFactory` in the `Jetty9Provider` class. 
we could change to something such (not tested :) )
```
Interface Provider

init(ServletContext svltCtxt)

```

`Websockets`

```
@MetaInfServices(ServletContainerInitializer.class)
public class WebSockets implements ServletContainerInitializer {

    public void onStartup(Set<Class<?>> set, ServletContext servletContext) throws ServletException {
        ServiceLoader.load(Provider.class).stream().forEach(service -> service.get().init(servletContext));
    }

```

`Jetty10Provider` 

```
   public void init(ServletContext servletContext) {
        JettyWebSocketServerContainer.ensureContainer(servletContext).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
    }
```
similar change for `Jetty9Provider`.


with this we do not store anymore anything in a field and do not use anymore some `synchronized` method

Reply:
> Can I assume you tested repeated agent connections within a single controller session?

Yes I did. Keeping a reference here is probably not needed. What @olamy proposes seems more consistent, to have an one time initializer on startup.

Reply:
Tested it in a separate branch https://github.com/jenkinsci/jenkins/commit/9ca7cb3e1f192b0d6c526fcd95c9108d7c85c9b0. And that seems to work just fine. I haven't tested for Jetty9 though.
@jglick let me know what you think. Using a `ServletContainerInitializer` looks good. Also how should I test the Jetty9 provider ? just remove the dependency on `websocket-jetty10` from the war/pom ?

Reply:
Since this PR is a fix for an important regression, which I have taken the liberty of marking `lts-candidate` for possible backport, let us keep the diff simple. Structural refactorings for stylistic purposes can be proposed later in trunk.

Reply:
Got it. So maybe I can use an `AtomicBoolean` to handle the single initialization and not cache the `JettyWebSocketServerContainer`.